src/network/gameclient.cc

```diff
@@ -651,6 +669,30 @@ void GameClient::handle_hello(RecvPacket& packet) {
 	}
 	d->settings.usernum = packet.unsigned_32();  // TODO(Klaus Halfmann): usernum is int8_t.
 	d->settings.playernum = -1;
+
+	d->addons_guard_.reset();
+	std::vector<AddOns::AddOnState> new_g_addons;
+	for (size_t i = packet.unsigned_32(); i; --i) {
+		const std::string name = packet.string();
+		const AddOns::AddOnVersion v = AddOns::string_to_version(packet.string());
+		AddOns::AddOnVersion found;
+		for (const auto& pair : AddOns::g_addons) {
+			if (pair.first.internal_name == name) {
+				found = pair.first.version;
+				new_g_addons.push_back(std::make_pair(pair.first, true));
+				break;
+			}
+		}
+		if (found.empty()) {
+			throw wexception("Add-on '%s' required by host not found", name.c_str());
+		}
+		if (found != v) {
+			throw wexception("Add-on '%s' installed at version %s but host uses version %s",
+			                 name.c_str(), AddOns::version_to_string(found, false).c_str(),
+			                 AddOns::version_to_string(v, false).c_str());
+		}
```

Comment:
I think an exception is a bit brutal here. At least the whole list should be checked to provide a complete list of missing addons. Currently only one is displayed and after it is installed the next error shows up. Also because of the wexception the file and line is displayed in the message box, but this has nothing to do with faulty code .